File: gem5/src/learning_gem5/part2/hello_object.cc
#include "learning_gem5/part2/hello_object.hh"


#include <iostream>
#include "base/logging.hh"
#include "debug/HelloExample.hh"

HelloObject::HelloObject(HelloObjectParams *params) :
    SimObject(params),
    // This is a C++ lambda. When the event is triggered, it will call the
    // processEvent() function. (this must be captured)
    event([this]{ processEvent(); }, name() + ".event"),
    goodbye(params->goodbye_object),
    // Note: This is not needed as you can *always* reference this->name()
    myName(params->name),
    latency(params->time_to_wait),
    timesLeft(params->number_of_fires)
{
    DPRINTF(HelloExample, "Created the hello object\n");
    panic_if(!goodbye, "HelloObject must have a non-null GoodbyeObject");
}

void
HelloObject::processEvent()
{
    timesLeft--;
    DPRINTF(HelloExample, "Process %d\n",timesLeft);

    if(timesLeft<=0){
        DPRINTF(HelloExample,"Finish!\n");
        goodbye->sayGoodbye(myName);
    }else{
        schedule(event,curTick()+latency);
    }
}

void
HelloObject::startup()
{
    schedule(event,latency);
}

HelloObject*
HelloObjectParams::create()
{
    return new HelloObject(this);
}

File: gem5/src/learning_gem5/part2/HelloObject.py
from m5.params import *
from m5.SimObject import SimObject

class HelloObject(SimObject):
    #It is not required that the type be the same as the name of the class, but it is convention. 
    type = 'HelloObject'
    cxx_header = "learning_gem5/part2/hello_object.hh"

    time_to_wait = Param.Latency("Time before firing the event")
    number_of_fires = Param.Int(1, "Number of times to fire the event before "
                                   "goodbye")

    goodbye_object = Param.GoodbyeObject("A goodbye object")


class GoodbyeObject(SimObject):
    type = 'GoodbyeObject'
    cxx_header = "learning_gem5/part2/goodbye_object.hh"

    buffer_size = Param.MemorySize('1kB',
                                   "Size of buffer to fill with goodbye")
    write_bandwidth = Param.MemoryBandwidth('100MB/s', "Bandwidth to fill "
                                            "the buffer")

    

    
